File: samples/Cpp/TestCpp/Classes/EffectsTest/EffectsTest.cpp
#include "EffectsTest.h"
#include "../testResource.h"

enum {
    kTagTextLayer = 1,

    kTagBackground = 1,
    kTagLabel = 2,
};

static int actionIdx=0; 
static std::string effectsList[] =
{
    "Shaky3D",
    "Waves3D",
    "FlipX3D",
    "FlipY3D",
    "Lens3D",
    "Ripple3D",
    "Liquid",
    "Waves",
    "Twirl",
    "ShakyTiles3D",
    "ShatteredTiles3D",
    "ShuffleTiles",
    "FadeOutTRTiles",
    "FadeOutBLTiles",
    "FadeOutUpTiles",
    "FadeOutDownTiles",
    "TurnOffTiles",
    "WavesTiles3D",
    "JumpTiles3D",
    "SplitRows",
    "SplitCols",
    "PageTurn3D",
}; 


class Shaky3DDemo : public Shaky3D 
{
public:
    static ActionInterval* create(float t)
    {
        return Shaky3D::create(t, Size(15,10), 5, false);
    }
};

class Waves3DDemo : public Waves3D 
{
public:
    static ActionInterval* create(float t)
    {
        return Waves3D::create(t, Size(15,10), 5, 40);
    }
};

class FlipX3DDemo : public FlipX3D 
{
public:
    static ActionInterval* create(float t)
    {
        FlipX3D* flipx  = FlipX3D::create(t);
        ActionInterval* flipx_back = flipx->reverse();
        DelayTime* delay = DelayTime::create(2);
        
        return Sequence::create(flipx, delay, flipx_back, NULL);
    }
};

class FlipY3DDemo : public FlipY3D 
{
public:
    static ActionInterval* create(float t)
    {
        FlipY3D* flipy  = FlipY3D::create(t);
        ActionInterval* flipy_back = flipy->reverse();
        DelayTime* delay = DelayTime::create(2);
        
        return Sequence::create(flipy, delay, flipy_back, NULL);
    }
};

class Lens3DDemo : public Lens3D 
{
public:
    static ActionInterval* create(float t)
    {
        Size size = Director::getInstance()->getWinSize();
        return Lens3D::create(t, Size(15,10), Point(size.width/2,size.height/2), 240); 
    }
};


class Ripple3DDemo : public Ripple3D 
{
public:
    static ActionInterval* create(float t)
    {
        Size size = Director::getInstance()->getWinSize();
        return Ripple3D::create(t, Size(32,24), Point(size.width/2,size.height/2), 240, 4, 160);
    }
};


class LiquidDemo : public Liquid
{
public:
    static ActionInterval* create(float t)
    {
        return Liquid::create(t, Size(16,12), 4, 20); 
    }
};


class WavesDemo : public Waves 
{
public:
    static ActionInterval* create(float t)
    {
        return Waves::create(t, Size(16,12), 4, 20, true, true);  
    }
};


class TwirlDemo : public Twirl 
{
public:
    static ActionInterval* create(float t)
    {
        Size size = Director::getInstance()->getWinSize();
        return Twirl::create(t, Size(12,8), Point(size.width/2, size.height/2), 1, 2.5f); 
    }
};


class ShakyTiles3DDemo : public ShakyTiles3D
{
public:
    static ActionInterval* create(float t)
    {
        return ShakyTiles3D::create(t, Size(16,12), 5, false) ;
    }
};


class ShatteredTiles3DDemo : public ShatteredTiles3D
{
public:
    static ActionInterval* create(float t)
    {
        return ShatteredTiles3D::create(t, Size(16,12), 5, false); 
    }
};


class ShuffleTilesDemo : public ShuffleTiles
{
public:
    static ActionInterval* create(float t)
    {
        ShuffleTiles* shuffle = ShuffleTiles::create(t, Size(16,12), 25);
        ActionInterval* shuffle_back = shuffle->reverse();
        DelayTime* delay = DelayTime::create(2);

        return Sequence::create(shuffle, delay, shuffle_back, NULL);
    }
};


class FadeOutTRTilesDemo : public FadeOutTRTiles
{
public:
    static ActionInterval* create(float t)
    {
        FadeOutTRTiles* fadeout = FadeOutTRTiles::create(t, Size(16,12));
        ActionInterval* back = fadeout->reverse();
        DelayTime* delay = DelayTime::create(0.5f);

        return Sequence::create(fadeout, delay, back, NULL);
    }
};


class FadeOutBLTilesDemo : public FadeOutBLTiles
{
public:
    static ActionInterval* create(float t)
    {
        FadeOutBLTiles* fadeout = FadeOutBLTiles::create(t, Size(16,12));
        ActionInterval* back = fadeout->reverse();
        DelayTime* delay = DelayTime::create(0.5f);

        return Sequence::create(fadeout, delay, back, NULL);
    }
};


class FadeOutUpTilesDemo : public FadeOutUpTiles
{
public:
    static ActionInterval* create(float t)
    {
        FadeOutUpTiles* fadeout = FadeOutUpTiles::create(t, Size(16,12));
        ActionInterval* back = fadeout->reverse();
        DelayTime* delay = DelayTime::create(0.5f);

        return Sequence::create(fadeout, delay, back, NULL);
    }
};

class FadeOutDownTilesDemo : public FadeOutDownTiles
{
public:
    static ActionInterval* create(float t)
    {
        FadeOutDownTiles* fadeout = FadeOutDownTiles::create(t, Size(16,12));
        ActionInterval* back = fadeout->reverse();
        DelayTime* delay = DelayTime::create(0.5f);

        return Sequence::create(fadeout, delay, back, NULL);
    }
};

class TurnOffTilesDemo : public TurnOffTiles
{
public:
    static ActionInterval* create(float t)
    {
        TurnOffTiles* fadeout = TurnOffTiles::create(t, Size(48,32), 25);
        ActionInterval* back = fadeout->reverse();
        DelayTime* delay = DelayTime::create(0.5f);

        return Sequence::create(fadeout, delay, back, NULL);
    }
};

class WavesTiles3DDemo : public WavesTiles3D
{
public:
    static ActionInterval* create(float t)
    {
        return WavesTiles3D::create(t, Size(15,10), 4, 120); 
    }
};

class JumpTiles3DDemo : public JumpTiles3D
{
public:
    static ActionInterval* create(float t)
    {
        return JumpTiles3D::create(t, Size(15,10), 2, 30); 
    }
};

class SplitRowsDemo : public SplitRows
{
public:
    static ActionInterval* create(float t)
    {
        return SplitRows::create(t, 9); 
    }
};

class SplitColsDemo : public SplitCols
{
public:
    static ActionInterval* create(float t)
    {
        return SplitCols::create(t, 9); 
    }
};

class PageTurn3DDemo : public PageTurn3D
{
public:
    static ActionInterval* create(float t)
    {
        Director::getInstance()->setDepthTest(true);
        return PageTurn3D::create(t, Size(15,10)); 
    }
};

//------------------------------------------------------------------
//
// TextLayer
//
//------------------------------------------------------------------
#define MAX_LAYER    22

ActionInterval* createEffect(int nIndex, float t)
{

    Director::getInstance()->setDepthTest(false);

    switch(nIndex)
    {
        case 0: return Shaky3DDemo::create(t);
        case 1: return Waves3DDemo::create(t);
        case 2: return FlipX3DDemo::create(t);
        case 3: return FlipY3DDemo::create(t);
        case 4: return Lens3DDemo::create(t);
        case 5: return Ripple3DDemo::create(t);
        case 6: return LiquidDemo::create(t);
        case 7: return WavesDemo::create(t);
        case 8: return TwirlDemo::create(t);
        case 9: return ShakyTiles3DDemo::create(t);
        case 10: return ShatteredTiles3DDemo::create(t);
        case 11: return ShuffleTilesDemo::create(t);
        case 12: return FadeOutTRTilesDemo::create(t);
        case 13: return FadeOutBLTilesDemo::create(t);
        case 14: return FadeOutUpTilesDemo::create(t);
        case 15: return FadeOutDownTilesDemo::create(t);
        case 16: return TurnOffTilesDemo::create(t);
        case 17: return WavesTiles3DDemo::create(t);
        case 18: return JumpTiles3DDemo::create(t);
        case 19: return SplitRowsDemo::create(t);
        case 20: return SplitColsDemo::create(t);
        case 21: return PageTurn3DDemo::create(t);
    }

    return NULL;
}

ActionInterval* getAction()
{
    ActionInterval* pEffect = createEffect(actionIdx, 3);

    return pEffect;
} 

void EffectTestScene::runThisTest()
{
    addChild(TextLayer::create());
    Director::getInstance()->replaceScene(this);
}

#define SID_RESTART        1

TextLayer::TextLayer(void)
: BaseTest()
{
	LayerColor *background = LayerColor::create( Color4B(32,128,32,255) );
	this->addChild(background,-20);
    
    Node* node = Node::create();
    ActionInterval* effect = getAction();
    node->runAction(effect);
    addChild(node, 0, kTagBackground);
    
    Sprite *bg = Sprite::create(s_back3);
    node->addChild(bg, 0);
//  bg->setAnchorPoint( Point::ZERO );
    bg->setPosition(VisibleRect::center());

    Sprite* grossini = Sprite::create(s_pathSister2);
    node->addChild(grossini, 1);
    grossini->setPosition( Point(VisibleRect::left().x+VisibleRect::getVisibleRect().size.width/3,VisibleRect::center().y) );
    ActionInterval* sc = ScaleBy::create(2, 5);
    ActionInterval* sc_back = sc->reverse();
    grossini->runAction( RepeatForever::create(Sequence::create(sc, sc_back, NULL) ) );

    Sprite* tamara = Sprite::create(s_pathSister1);
    node->addChild(tamara, 1);
    tamara->setPosition( Point(VisibleRect::left().x+2*VisibleRect::getVisibleRect().size.width/3,VisibleRect::center().y) );
    ActionInterval* sc2 = ScaleBy::create(2, 5);
    ActionInterval* sc2_back = sc2->reverse();
    tamara->runAction( RepeatForever::create(Sequence::create(sc2, sc2_back, NULL)) );
    
    LabelTTF* label = LabelTTF::create((effectsList[actionIdx]).c_str(), "Marker Felt", 32);
    
    label->setPosition( Point(VisibleRect::center().x,VisibleRect::top().y-80) );
    addChild(label);
    label->setTag( kTagLabel );
    
    schedule( schedule_selector(TextLayer::checkAnim) );
}

void TextLayer::checkAnim(float dt)
{
    Node* s2 = getChildByTag(kTagBackground);
    if ( s2->getNumberOfRunningActions() == 0 && s2->getGrid() != NULL)
        s2->setGrid(NULL);;
}


TextLayer::~TextLayer(void)
{
}

// TextLayer* TextLayer::node()
// {
//     return TextLayer::create();
// }

TextLayer* TextLayer::create()
{
    TextLayer* layer = new TextLayer();
    layer->autorelease();
    
    return layer;
}

void TextLayer::onEnter()
{
    BaseTest::onEnter();
}

void TextLayer::newScene()
{
    Scene* s = new EffectTestScene();
    Node* child = TextLayer::create();
    s->addChild(child);
    Director::getInstance()->replaceScene(s);
    s->release();
}

void TextLayer::restartCallback(Object* sender)
{
    newScene();
}

void TextLayer::nextCallback(Object* sender)
{
    // update the action index
    actionIdx++;
    actionIdx = actionIdx % MAX_LAYER;

    newScene();
}

void TextLayer::backCallback(Object* sender)
{
    // update the action index
    actionIdx--;
    int total = MAX_LAYER;
    if( actionIdx < 0 )
        actionIdx += total;    

    newScene();
} 

File: tests/test-lua/Resources/luaScript/EffectsTest/EffectsTest.lua
require "luaScript/EffectsTest/EffectsName"


local ActionIdx = -1
local size = cc.Director:getInstance():getWinSize()
local kTagTextLayer  = 1
local kTagBackground = 1
local kTagLabel      = 2
local MAX_LAYER      = 22

local testLayer  = nil
local titleLabel = nil
local entry      = nil
local gridNodeTarget = nil

local function checkAnim(dt)

    if nil == gridNodeTarget then
        return
    end

    if gridNodeTarget:getNumberOfRunningActions() == 0 and nil ~=gridNodeTarget:getGrid() then
        gridNodeTarget:setGrid(nil)
    end
end

local function onEnterOrExit(tag)
    local scheduler = cc.Director:getInstance():getScheduler()
    if tag == "enter" then
        entry = scheduler:scheduleScriptFunc(checkAnim, 0, false)
    elseif tag == "exit" then
        scheduler:unscheduleScriptEntry(entry)
    end
end

local function backAction()
    ActionIdx = ActionIdx - 1
    if ActionIdx < 0 then
        ActionIdx = ActionIdx + MAX_LAYER
    end

    return CreateEffectsTestLayer()
end

local function restartAction()
    return CreateEffectsTestLayer()
end

local function nextAction()
    ActionIdx = ActionIdx + 1
    ActionIdx = math.mod(ActionIdx, MAX_LAYER)

    return CreateEffectsTestLayer()
end

local function backCallback(sender)
    local scene = cc.Scene:create()

    scene:addChild(backAction())
    scene:addChild(CreateBackMenuItem())

    cc.Director:getInstance():replaceScene(scene)
end

local function restartCallback(sender)
    local scene = cc.Scene:create()

    scene:addChild(restartAction())
    scene:addChild(CreateBackMenuItem())

    cc.Director:getInstance():replaceScene(scene)
end

local function nextCallback(sender)
    local scene = cc.Scene:create()

    scene:addChild(nextAction())
    scene:addChild(CreateBackMenuItem())

    cc.Director:getInstance():replaceScene(scene)
end

--------------------------------------
--  Shaky3DDemo
--------------------------------------
local function Shaky3DDemo(t)
    return cc.Shaky3D:create(t, cc.size(15,10), 5, false)
end

--------------------------------------
--  Waves3DDemo
--------------------------------------
local function Waves3DDemo(t)
    return cc.Waves3D:create(t, cc.size(15,10), 5, 40)
end

--------------------------------------
--  FlipX3DDemo
--------------------------------------
local function FlipX3DDemo(t)
    local flipx  = cc.FlipX3D:create(t)
    local flipx_back = flipx:reverse()
    local delay = cc.DelayTime:create(2)

    return cc.Sequence:create(flipx, flipx_back, delay)
end

--------------------------------------
--  FlipY3DDemo
--------------------------------------
local function FlipY3DDemo(t)
    local flipy  = cc.FlipY3D:create(t)
    local flipy_back = flipy:reverse()
    local delay = cc.DelayTime:create(2)

    return cc.Sequence:create(flipy, flipy_back, delay)
end

--------------------------------------
--  Lens3DDemo
--------------------------------------
local function Lens3DDemo(t)
    return cc.Lens3D:create(t, cc.size(15,10), cc.p(size.width/2,size.height/2), 240)
end

--------------------------------------
--  Ripple3DDemo
--------------------------------------
local function Ripple3DDemo(t)
    return cc.Ripple3D:create(t, cc.size(32,24), cc.p(size.width/2,size.height/2), 240, 4, 160)
end

--------------------------------------
--  LiquidDemo
--------------------------------------
local function LiquidDemo(t)
    return cc.Liquid:create(t, cc.size(16,12), 4, 20)
end

--------------------------------------
--  WavesDemo
--------------------------------------
local function WavesDemo(t)
    return cc.Waves:create(t, cc.size(16,12), 4, 20, true, true)
end

--------------------------------------
--  TwirlDemo
--------------------------------------
local function TwirlDemo(t)
    return cc.Twirl:create(t, cc.size(12,8), cc.p(size.width/2, size.height/2), 1, 2.5)
end

--------------------------------------
--  ShakyTiles3DDemo
--------------------------------------
local function ShakyTiles3DDemo(t)
    return cc.ShakyTiles3D:create(t, cc.size(16,12), 5, false)
end

--------------------------------------
--  ShatteredTiles3DDemo
--------------------------------------
local function ShatteredTiles3DDemo(t)
    return cc.ShatteredTiles3D:create(t, cc.size(16,12), 5, false)
end

--------------------------------------
--  ShuffleTilesDemo
--------------------------------------
local function ShuffleTilesDemo(t)
    local shuffle = cc.ShuffleTiles:create(t, cc.size(16,12), 25)
    local shuffle_back = shuffle:reverse()
    local delay = cc.DelayTime:create(2)

    return cc.Sequence:create(shuffle, shuffle_back, delay)
end

--------------------------------------
--  FadeOutTRTilesDemo
--------------------------------------
local function FadeOutTRTilesDemo(t)
    local fadeout = cc.FadeOutTRTiles:create(t, cc.size(16,12))
    local back = fadeout:reverse()
    local delay = cc.DelayTime:create(0.5)

    return cc.Sequence:create(fadeout, back, delay)
end

--------------------------------------
--  FadeOutBLTilesDemo
--------------------------------------
local function FadeOutBLTilesDemo(t)
    local fadeout = cc.FadeOutBLTiles:create(t, cc.size(16,12))
    local back = fadeout:reverse()
    local delay = cc.DelayTime:create(0.5)

    return cc.Sequence:create(fadeout, back, delay)
end

--------------------------------------
--  FadeOutUpTilesDemo
--------------------------------------
local function FadeOutUpTilesDemo(t)
    local fadeout = cc.FadeOutUpTiles:create(t, cc.size(16,12))
    local back = fadeout:reverse()
    local delay = cc.DelayTime:create(0.5)

    return cc.Sequence:create(fadeout, back, delay)
end

--------------------------------------
--  FadeOutDownTilesDemo
--------------------------------------
local function FadeOutDownTilesDemo(t)
    local fadeout = cc.FadeOutDownTiles:create(t, cc.size(16,12))
    local back = fadeout:reverse()
    local delay = cc.DelayTime:create(0.5)

    return cc.Sequence:create(fadeout, back, delay)
end

--------------------------------------
--  TurnOffTilesDemo
--------------------------------------
local function TurnOffTilesDemo(t)
    local fadeout = cc.TurnOffTiles:create(t, cc.size(48,32), 25)
    local back = fadeout:reverse()
    local delay = cc.DelayTime:create(0.5)

    return cc.Sequence:create(fadeout, back, delay)
end

--------------------------------------
--  WavesTiles3DDemo
--------------------------------------
local function WavesTiles3DDemo(t)
    return cc.WavesTiles3D:create(t, cc.size(15,10), 4, 120)
end

--------------------------------------
--  JumpTiles3DDemo
--------------------------------------
local function JumpTiles3DDemo(t)
    return cc.JumpTiles3D:create(t, cc.size(15,10), 2, 30)
end

--------------------------------------
--  SplitRowsDemo
--------------------------------------
local function SplitRowsDemo(t)
    return cc.SplitRows:create(t, 9)
end

--------------------------------------
--  SplitColsDemo
--------------------------------------
local function SplitColsDemo(t)
    return cc.SplitCols:create(t, 9)
end

--------------------------------------
--  PageTurn3DDemo
--------------------------------------
local function PageTurn3DDemo(t)
    cc.Director:getInstance():setDepthTest(true)
    return cc.PageTurn3D:create(t, cc.size(15,10))
end

--------------------------------------
--  Effects Test
--------------------------------------
local function createEffect(idx, t)
    cc.Director:getInstance():setDepthTest(false)
    local action = nil

    if idx == 0 then
        action = Shaky3DDemo(t)
    elseif idx == 1 then
        action = Waves3DDemo(t)
    elseif idx == 2 then
        action = FlipX3DDemo(t)
    elseif idx == 3 then
        action = FlipY3DDemo(t)
    elseif idx == 4 then
        action = Lens3DDemo(t)
    elseif idx == 5 then
        action = Ripple3DDemo(t)
    elseif idx == 6 then
        action = LiquidDemo(t)
    elseif idx == 7 then
        action = WavesDemo(t)
    elseif idx == 8 then
        action = TwirlDemo(t)
    elseif idx == 9 then
        action = ShakyTiles3DDemo(t)
    elseif idx == 10 then
        action = ShatteredTiles3DDemo(t)
    elseif idx == 11 then
        action = ShuffleTilesDemo(t)
    elseif idx == 12 then
        action = FadeOutTRTilesDemo(t)
    elseif idx == 13 then
        action = FadeOutBLTilesDemo(t)
    elseif idx == 14 then
        action = FadeOutUpTilesDemo(t)
    elseif idx == 15 then
        action = FadeOutDownTilesDemo(t)
    elseif idx == 16 then
        action = TurnOffTilesDemo(t)
    elseif idx == 17 then
        action = WavesTiles3DDemo(t)
    elseif idx == 18 then
        action = JumpTiles3DDemo(t)
    elseif idx == 19 then
        action = SplitRowsDemo(t)
    elseif idx == 20 then
        action = SplitColsDemo(t)
    elseif idx == 21 then
        action = PageTurn3DDemo(t)
    end

    return action
end

function CreateEffectsTestLayer()
    testLayer = cc.LayerColor:create(cc.c4b(32,128,32,255))


    gridNodeTarget = cc.NodeGrid:create()
    local effect = createEffect(ActionIdx, 3)
    gridNodeTarget:runAction(effect)
    testLayer:addChild(gridNodeTarget, 0, kTagBackground)
    
    local bg = cc.Sprite:create(s_back3)
    gridNodeTarget:addChild(bg, 0)
    bg:setPosition(VisibleRect:center())

    local grossini = cc.Sprite:create(s_pPathSister2)
    gridNodeTarget:addChild(grossini, 1)
    grossini:setPosition( cc.p(VisibleRect:left().x+VisibleRect:getVisibleRect().width/3,VisibleRect:center().y) )
    local sc = cc.ScaleBy:create(2, 5)
    local sc_back = sc:reverse()
    grossini:runAction( cc.RepeatForever:create(cc.Sequence:create(sc, sc_back) ) )

    local tamara = cc.Sprite:create(s_pPathSister1)
    gridNodeTarget:addChild(tamara, 1)
    tamara:setPosition( cc.p(VisibleRect:left().x+2*VisibleRect:getVisibleRect().width/3,VisibleRect:center().y) )
    local sc2 = cc.ScaleBy:create(2, 5)
    local sc2_back = sc2:reverse()
    tamara:runAction( cc.RepeatForever:create(cc.Sequence:create(sc2, sc2_back)) )

    local x, y = size.width, size.height

    titleLabel = cc.LabelTTF:create(EffectsList[ActionIdx], "Marker Felt", 32)
    titleLabel:setPosition(x / 2, y - 80)
    testLayer:addChild(titleLabel)
    titleLabel:setTag(kTagLabel)

    local item1 = cc.MenuItemImage:create(s_pPathB1, s_pPathB2)
    local item2 = cc.MenuItemImage:create(s_pPathR1, s_pPathR2)
    local item3 = cc.MenuItemImage:create(s_pPathF1, s_pPathF2)
    item1:registerScriptTapHandler(backCallback)
    item2:registerScriptTapHandler(restartCallback)
    item3:registerScriptTapHandler(nextCallback)

    local menu = cc.Menu:create()
    menu:addChild(item1)
    menu:addChild(item2)
    menu:addChild(item3)

    menu:setPosition(cc.p(0, 0))
    item1:setPosition(cc.p(size.width/2 - item2:getContentSize().width * 2, item2:getContentSize().height / 2))
    item2:setPosition(cc.p(size.width/2, item2:getContentSize().height / 2))
    item3:setPosition(cc.p(size.width/2 + item2:getContentSize().width * 2, item2:getContentSize().height / 2))

    testLayer:addChild(menu, 1)

    testLayer:registerScriptHandler(onEnterOrExit)

    return testLayer
end

function EffectsTest()
    local scene = cc.Scene:create()

    ActionIdx = -1
    scene:addChild(nextAction())
    scene:addChild(CreateBackMenuItem())

    return scene
end
